tests/middleware/test_base.py

```diff
@@ -206,3 +213,81 @@ async def homepage(request):
     client = test_client_factory(app)
     response = client.get("/")
     assert response.status_code == 200, response.content
+
+
+@pytest.mark.anyio
+async def test_background_tasks_client_disconnect() -> None:
+    # test for https://github.com/encode/starlette/issues/1438
+    container: List[str] = []
+
+    disconnected = anyio.Event()
+
+    async def slow_background() -> None:
+        await disconnected.wait()
+        container.append("called")
```

Comment:
I do not think that the background task itself should wait on the `disconnected` event - that places an explicit link between the ASGI events of the test and the background task, whereas the link that we seek to test (around cancellation and request processing in the presence of a `BaseHTTPMiddleware`) is implicit.

I think a `sleep(0.1)` to allow cancellation would be more appropriate.

```suggestion
    async def slow_background() -> None:
        # small delay to give BaseHTTPMiddleware a chance to cancel us
        # this is required to make the test fail prior to fixing the issue
        # so do not be surprised if you remove it and the test still passes
        await anyio.sleep(0.1)
        container.append("called")
```